web/src/drafts.js

```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
 
 export const DRAFT_LIFETIME = 30;
 
+export function get_draft_ids_for_topic(stream_id, topic) {
+    const drafts = draft_model.get();
+    return Object.keys(drafts)
+        .filter(
+            (draft_id) =>
+                drafts[draft_id].type === "stream" &&
+                drafts[draft_id].stream_id === stream_id &&
+                drafts[draft_id].topic === topic,
+        )
+        .sort((draft_a, draft_b) => drafts[draft_a].updatedAt - drafts[draft_b].updatedAt);
+}
+
+export function get_draft_ids_for_private_message_recipient(private_message_recipient) {
+    const drafts = draft_model.get();
+    return Object.keys(drafts)
+        .filter(
+            (draft_id) =>
+                drafts[draft_id].type === "private" &&
+                _.isEqual(
+                    drafts[draft_id].private_message_recipient.split(",").sort(),
+                    private_message_recipient.split(",").sort(),
+                ),
+        )
+        .sort((draft_a, draft_b) => drafts[draft_a].updatedAt - drafts[draft_b].updatedAt);
+}
+
+export function restore_last_draft_based_on_compose_state() {
+    const message_type = compose_state.get_message_type();
+    if (!message_type || compose_state.message_content()) {
+        return;
+    }
+
+    let last_draft_id;
+    if (message_type === "private") {
+        const recipient = compose_state.private_message_recipient();
+        if (recipient) {
+            last_draft_id = get_draft_ids_for_private_message_recipient(recipient).pop();
+        }
+    } else {
+        const stream_id = compose_state.stream_id();
+        const topic = compose_state.topic();
+        if (stream_id !== undefined && topic !== "") {
+            last_draft_id = get_draft_ids_for_topic(stream_id, topic).pop();
+        }
+    }
+
+    const draft = draft_model.getDraft(last_draft_id);
+    if (!draft) {
+        return;
+    }
+
+    compose_state.message_content(draft.content);
+    compose_ui.autosize_textarea($("#compose-textarea"));
+    $("#compose-textarea").trigger("select").data("draft-id", last_draft_id);
```

Comment:
What do you think about having this function just return a `draft_id` or undefined, and the compose code can then add that `draft_id` to `opts` and process accordingly?

That seems like it'd be much more obviously producing an equivalent result to having restored the draft another way, and would let us remove these violations of the `compose_actions` abstraction boundary that it's in charge of adjusting the compose UI.

Reply:
That sounds like a great idea, thanks. I hadn't considered that quote reply and whatnot would overwrite restored drafts, so I'm really glad you thought of that.